src/mozanalysis/metrics/desktop.py

```diff
@@ -2,95 +2,84 @@
-def active_hours(cd):
-    """Active hours, from ``active_ticks``
-
-    At any given moment, a client is "active" if there was a keyboard or
-    mouse interaction (click, scroll, move) in the previous 5 seconds.
-    """
-    return agg_sum(cd.active_hours_sum)
-
-
-@Metric.from_func(search_clients_daily)
-def search_count(scd):
-    return agg_sum(scd.sap)
-
-
-@Metric.from_func(search_clients_daily)
-def ad_clicks(scd):
-    return agg_sum(scd.ad_click)
-
-
-@Metric.from_func(clients_daily)
-def uri_count(cd):
-    return agg_sum(cd.scalar_parent_browser_engagement_total_uri_count_sum)
-
-
-@Metric.from_func(events)
-def unenroll(events, experiment):
-    return agg_any(all_([
-        events.event_category == 'normandy',
-        events.event_method == 'unenroll',
-        events.event_string_value == experiment.experiment_slug,
-    ]))
-
-
-@Metric.from_func(telemetry_shield_study_parquet)
-def unenroll_addon_expt(tssp):
-    return agg_any(tssp.payload.data.study_state == 'exit')
-
-
-@Metric.from_func(search_clients_daily)
-def organic_search_count(scd):
-    return agg_sum(scd.organic)
-
-
-@Metric.from_func(events)
-def view_about_logins(events):
-    return agg_any(all_([
-        events.event_method == 'open_management',
-        events.event_category == 'pwmgr',
-    ]))
 
 
-@Metric.from_func(events)
-def view_about_protections(events):
-    return agg_any(all_([
-        events.event_object == 'protection_report',
-        events.event_method == 'show',
-    ]))
+clients_daily = DataSource(
+    name='clients_daily',
+    from_expr="`moz-fx-data-shared-prod.telemetry.clients_daily`",
+)
+
+search_clients_daily = DataSource(
+    name='search_clients_daily',
+    from_expr='`moz-fx-data-shared-prod.search.search_clients_daily`',
+    experiments_column_type=None,
+)
+
+main_summary = DataSource(
+    name='main_summary',
+    from_expr="`moz-fx-data-shared-prod.telemetry.main_summary`"
+)
+
+events = DataSource(
+    name='events',
+    from_expr="`moz-fx-data-shared-prod.telemetry.events`"
+)
+
+active_hours = Metric(
+    name='active_hours',
+    data_source=clients_daily,
+    select_expr=agg_sum('active_hours_sum')
+)
+
+uri_count = Metric(
+    name='uri_count',
+    data_source=clients_daily,
+    select_expr=agg_sum('scalar_parent_browser_engagement_total_uri_count_sum')
+)
+
+search_count = Metric(
+    name='search_count',
+    data_source=search_clients_daily,
+    select_expr=agg_sum('sap')
+)
+
+ad_clicks = Metric(
+    name='ad_clicks',
+    data_source=search_clients_daily,
+    select_expr=agg_sum('ad_click')
+)
+
+organic_search_count = Metric(
+    name='organic_search_count',
+    data_source=search_clients_daily,
+    select_expr=agg_sum('organic')
+)
+
+unenroll = Metric(
+    name='unenroll',
+    data_source=events,
+    select_expr=agg_any("""
+            event_category = 'normandy'
+            AND event_method = 'unenroll'
+            AND event_string_value = '{experiment_slug}'
```

Comment:
What are the rules for templating?

Reply:
`{experiment_slug}` is provided; IIRC nothing else is.

If you get stuck and need other templating, then you can define your metrics on the fly, scratching your own itch. If this is a common occurrence then we can add extra templating as required, safe in the knowledge that we're not blocking anyone in the meantime

Reply:
Okay; please document that on the Metric class.